lib/iris/cube.py

```diff
@@ -2263,35 +2263,125 @@ def _summary_extra(self, coords, summary, indent):
-                dim_names[dim].add(dim_coords[0].name())
-            else:
-                dim_names[dim].add("-- ")
+        if len(dim_coords) == 1:
+            # Only 1 dimension coordinate can map to a dimension.
+            name = dim_coords[0].name()  # N.B. can only be 1
+        elif len(dim_coords) == 0:
+            # Anonymous dimension : NOTE this has an extra space on the end.
+            name = "-- "
+        else:
+            # >1 matching dim coord : This really can't happen !
+            names_list = [co.name() for co in dim_coords]
+            msg = (
+                "Cube logic error : "
+                "multiple dimension coords map to "
+                "dimension {dim} : {names!r}"
+            )
+            raise ValueError(msg.format(dim=dim, names=names_list))
+
+        return name
+
+    _VectorSectionSpec = namedtuple(
+        "Spec", ("title", "elements", "add_extra_lines")
+    )
+
+    def _summary_vector_sections_info(self):
+        """
+        Return a table describing the order and content of the 'vector'
+        sections of the cube summary.
+
+        Returns:
+
+            * summary_info (list of cube._VectorSectionSpec)
+                A list of summary sections info, in summary order.
+
+        The table entries contain arguments to be passed to the
+        'vector_summary' inner function, inside cube.summary().
+
+        Note: provided as a separate function so that derived cube types
+        (as in iris-ugrid) can override / extend the cube summary.
+
+        """
+        # Cache the derived coords so we can rely on consistent
+        # object IDs.
+        derived_coords = self.derived_coords
+        # Determine the cube coordinates that are scalar (single-valued)
+        # AND non-dimensioned.
+        dim_coords = self.dim_coords
+        aux_coords = self.aux_coords
+        all_coords = dim_coords + aux_coords + derived_coords
+        scalar_coords = [
+            coord
+            for coord in all_coords
+            if not self.coord_dims(coord) and coord.shape == (1,)
+        ]
+        # Determine the cube coordinates that are not scalar BUT
+        # dimensioned.
+        scalar_coord_ids = set(map(id, scalar_coords))
+        vector_dim_coords = [
+            coord for coord in dim_coords if id(coord) not in scalar_coord_ids
+        ]
+        vector_aux_coords = [
+            coord for coord in aux_coords if id(coord) not in scalar_coord_ids
+        ]
+        vector_derived_coords = [
+            coord
+            for coord in derived_coords
+            if id(coord) not in scalar_coord_ids
+        ]
+
+        # cell measures
+        vector_cell_measures = [
+            cm for cm in self.cell_measures() if cm.shape != (1,)
+        ]
+
+        # Ancillary Variables
+        vector_ancillary_variables = [av for av in self.ancillary_variables()]
+
+        # Sort scalar coordinates by name.
+        scalar_coords.sort(key=lambda coord: coord.name())
```

Comment:
Good spot !

Actually it's worse than that .. this variable is never used again (it's local to the routine).
To avoid a nastier API to the routine, I chose to re-calculate this separately in the main routine.
So in _this_ routine we actually only use the 'ids'. I will tidy that up.